### PR TITLE
feat: Code block

### PR DESCRIPTION
**Pull request by areknawo**
This feature was sponsored by [CellaJS](https://github.com/cellajs/cella) 💖

This PR adds a code block with syntax highlighting (using [prosemirror-highlight](https://github.com/ocavue/prosemirror-highlight)) to the core.

BlockNote Code block Example

closes https://github.com/TypeCellOS/BlockNote/issues/721
closes https://github.com/TypeCellOS/BlockNote/issues/147
closes https://github.com/TypeCellOS/BlockNote/pull/1099
closes https://github.com/TypeCellOS/BlockNote/issues/721

### CODE REVIEW

**Review by YousefED — COMMENTED** [review]
This looks great!

I think the only things remaining would be to add this block to the nodeconversions / htmlconversions unit tests (we can also do this). 

I also noticed that we don't correctly set the language when it's parsed from markdown (you can try this here: https://blocknote-git-fork-areknawo-main-typecell.vercel.app/interoperability/converting-blocks-from-md)

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+  content: "inline*",
+  group: "blockContent",
+  marks: "",
+  code: true,
+  defining: true,
+  addOptions() {
+    return {
+      indentLineWithTab: true,
+      supportedLanguages: [
+        {
```
[the remark this answers is not in the document]
Comment (areknawo) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:44]: I also wasn't sure about this. Usually, in HTML/MD, code blocks without language simply leave it out, so I made an empty string the default and assigned a *Code* label to it. I think the empty string makes sense, but label can be changed

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+      Delete: ({ editor }) => {
+        const { selection } = editor.state;
+        const { $from } = selection;
+
+        if (
+          editor.isActive(this.name) &&
+          !$from.parent.textContent &&
+          isTextSelection(selection)
+        ) {
+          const from = $from.pos - $from.parentOffset - 2;
```
[the remark this answers is not in the document]
Comment (areknawo) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:279]: Do you mean a specific line or the keyboard shortcut as a whole?

As for the shortcut - this one removes an empty code block on `DELETE` from the inside (common behavior across different WYSIWYG editors that didn't work out-of-the-box in BlockNote). I can add a comment

File: playground/package.json [hunk 3]
```diff
@@ -3,7 +3,7 @@
   "private": true,
   "version": "0.17.1",
   "scripts": {
-    "dev": "vite",
+    "dev": "vite --host",
```
[the remark this answers is not in the document]
Comment (areknawo) [reply playground/package.json:6]: It exposes the dev server on the local network. I didn't mean to leave it, but it's helpful for testing on different devices

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+import { createDefaultBlockDOMOutputSpec } from "../defaultBlockHelpers.js";
+
+type SupportedLanguageConfig = {
+  id: string;
+  name: string;
+  match: string[];
+};
+
+export const codeBlockPropSchema = {
+  ...defaultProps,
```
Comment (areknawo) [inline packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:26]: @YousefED Should the code block support default props (e.g. `textAlignment`, `backgroundColor`, etc.). I've referenced it from other blocks, but I'm not sure if it makes sense here

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (YousefED) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:44]: I think we should have a `text` language (https://shiki.style/languages#plain-text) for plain text.

Then we can still decide to upon creation initialize with plain text vs javascript. Maybe let's do javascript for now because otherwise people don't really notice the syntax highlighting

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (YousefED) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:279]: The specific line. My reasoning; when I read this code with a manual position fix in the review, I figured there's probably something interesting going on - so I was triggered to ask about this. From your reply that seems to be the case indeed, so let's document it with a comment and turn this into shared knowledge.

TLDR: if there's a "workaround" needed for something, let's try to document that as much as possible in the code for others or your "future self"

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts (hunk 4 above)
Comment (YousefED) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:26]: good spot, let's remove

**Comment by YousefED** [comment]
> I see that you specifically modify rehype handler to use a data-language attribute. I can add support for it in the block.
As for the handler - it also adds a newline at the end of the block - what's the purpose of it and can it be removed?

Cool. newline, not sure, let's remove it (might have been TypeCell related)

> Makes sense to have a smaller list

What about this list: https://shiki.style/languages -> web bundle, but 
- without: angular, astro, blade, coffee, handlebars, html derivative, http, imba, jinja, jison, json5, marko, mdc, stylus, ts-tags 
- with: haskell, c#, latex, lua, mermaid, ruby, rust, scala, swift, kotlin, objective c

What do you think?

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (areknawo) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:44]: Ok, I think I can add `defaultLanguage` option with `text` as the default.
For `text`, `txt`, `plain` values, I understand `renderHTML` should skip them, not to add `class="language-text"` to the HTML?

**Comment by areknawo** [comment]
> What do you think?

I think it's a good selection. It's always going to be a bit subjective, but it makes sense to shorten it a little bit by default

**Review by YousefED — APPROVED** [review]

File: packages/core/src/api/parsers/markdown/parseMarkdown.ts [hunk 5]
```diff
@@ -12,7 +12,7 @@ import { HTMLToBlocks } from "../html/parseHTML.js";
 // modified version of https://github.com/syntax-tree/mdast-util-to-hast/blob/main/lib/handlers/code.js
 // that outputs a data-language attribute instead of a CSS class (e.g.: language-typescript)
 function code(state: any, node: any) {
-  const value = node.value ? node.value + "\n" : "";
+  const value = node.value ? node.value : "";
```
[the remark this answers is not in the document]
Comment (YousefED) [reply packages/core/src/api/parsers/markdown/parseMarkdown.ts:15]: previous part was I think a typecell-specific fix. At least, it was not documented so better to remove until we need it again

File: packages/core/src/api/clipboard/fromClipboard/handleVSCodePaste.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+>(event: ClipboardEvent, editor: BlockNoteEditor<BSchema, I, S>) {
+  const view = editor.prosemirrorView;
+  const { schema } = view.state;
+
+  if (!event.clipboardData) {
+    return false;
+  }
+
+  const text = event.clipboardData!.getData("text/plain");
+  const vscode = event.clipboardData!.getData("vscode-editor-data");
```
[the remark this answers is not in the document]
Comment (YousefED) [reply packages/core/src/api/clipboard/fromClipboard/handleVSCodePaste.ts:21]: fixed in #1219

File: packages/core/src/extensions/SuggestionMenu/getDefaultSlashMenuItems.ts [hunk 7]
```diff
@@ -175,6 +175,24 @@ export function getDefaultSlashMenuItems<
...
+    items.push({
+      onItemClick: () => {
+        const pos = editor._tiptapEditor.state.selection.from;
+
+        insertOrUpdateBlock(editor, {
+          type: "codeBlock",
+        });
+
+        // Move the cursor inside the code block
+        editor._tiptapEditor.commands.setTextSelection(pos);
```
[the remark this answers is not in the document]
Comment (YousefED) [reply packages/core/src/extensions/SuggestionMenu/getDefaultSlashMenuItems.ts:188]: fixed in #1219

File: packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    };
+  },
+});
+
+export const CodeBlock = createBlockSpecFromStronglyTypedTiptapNode(
+  CodeBlockContent,
+  defaultCodeBlockPropSchema
+);
+
+export function customizeCodeBlock(options: Partial<CodeBlockOptions>) {
```
[the remark this answers is not in the document]
Comment (YousefED) [reply packages/core/src/blocks/CodeBlockContent/CodeBlockContent.ts:355]: agree, but let's check when we create docs for this
